samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -87,21 +275,75 @@ def _process_reference_resource_value(self, resolved_destination_resource: Resol
+            )
 
-        Parameters
-        ----------
-        resolved_destination_resource: List[Union[ConstantValue, ResolvedReference]]
-            The resolved destination resources to be processed for the input source resource.
+        # Valid destination resource
+        if resolved_destination_resource.value.startswith(self._resource_pair.terraform_resource_type_prefix):
+            LOG.debug("Process the destination resource %s", resolved_destination_resource.value)
+            if not resolved_destination_resource.value.endswith("arn"):
+                LOG.debug(
+                    "The used property in reference %s is not an ARN property", resolved_destination_resource.value
+                )
+                raise InvalidResourceLinkingException(
+                    f"Could not use the value {resolved_destination_resource.value} as a "
+                    f"destination resource for the source resource "
+                    f"{source_tf_resource.full_address}. The source resource "
+                    f"value should refer to valid destination resource ARN property."
+                )
 
-        Returns
-        --------
-        List[Dict[str, str]]:
-            The list of destination resources after processing
-        """
+            tf_dest_res_name = resolved_destination_resource.value[
+                len(self._resource_pair.terraform_resource_type_prefix) : -len(".arn")
+            ]
+            if resolved_destination_resource.module_address:
+                tf_dest_resource_full_address = (
+                    f"{resolved_destination_resource.module_address}."
+                    f"{self._resource_pair.terraform_resource_type_prefix}"
+                    f"{tf_dest_res_name}"
+                )
+            else:
+                tf_dest_resource_full_address = (
+                    f"{self._resource_pair.terraform_resource_type_prefix}" f"{tf_dest_res_name}"
+                )
+            cfn_dest_resource_logical_id = build_cfn_logical_id(tf_dest_resource_full_address)
+            LOG.debug(
+                "The logical id of the resource referred by %s is %s",
+                resolved_destination_resource.value,
+                cfn_dest_resource_logical_id,
+            )
+
+            # validate that the found dest resource is in mapped dest resources, which means that it is created.
+            # The resource can be defined in the TF plan configuration, but will not be created.
+            dest_resources = []
+            if cfn_dest_resource_logical_id in self._resource_pair.destination_resource_tf:
+                LOG.debug(
+                    "The resource referred by %s can be found in the mapped destination resources",
+                    resolved_destination_resource.value,
+                )
+                dest_resources.append({self._resource_pair.intrinsic_type.value: cfn_dest_resource_logical_id})
```

Comment:
replace `self._resource_pair.intrinsic_type.value` with "Ref" for now. If it is `GetAtt`, we need to set a list that contains the logical id and the attribute name